server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/WriteLoadConstraintDeciderTests.java

```diff
@@ -247,6 +268,106 @@ public void testWriteLoadDeciderCanRemain() {
+                Map.of(
+                    overloadedNode.getId(),
+                    new NodeUsageStatsForThreadPools(
+                        overloadedNode.getId(),
+                        Map.of(
+                            ThreadPool.Names.WRITE,
+                            new ThreadPoolUsageStats(
+                                numThreads,
+                                randomFloatBetween(highUtilizationThreshold, 1.1f, false),
+                                randomLongBetween(highLatencyThreshold, highLatencyThreshold * 2)
+                            )
+                        )
+                    ),
+                    otherNode.getId(),
+                    new NodeUsageStatsForThreadPools(
+                        otherNode.getId(),
+                        Map.of(
+                            ThreadPool.Names.WRITE,
+                            new ThreadPoolUsageStats(
+                                numThreads,
+                                randomFloatBetween(0.0f, highUtilizationThreshold / 2, true),
+                                randomLongBetween(0, highLatencyThreshold / 2)
+                            )
+                        )
+                    )
+                )
+            )
+            // simulate all zero or missing shard write loads
+            .shardWriteLoads(
+                state.routingTable(ProjectId.DEFAULT).allShards().collect(Collectors.toMap(ShardRouting::shardId, ignored -> 0.0d))
+            )
+            .build();
+
+        final var clusterSettings = createBuiltInClusterSettings(settings);
+        final var writeLoadConstraintDecider = new WriteLoadConstraintDecider(clusterSettings);
+        final var routingAllocation = new RoutingAllocation(
+            new AllocationDeciders(List.of(writeLoadConstraintDecider)),
+            state.getRoutingNodes().mutableCopy(),
+            state,
+            clusterInfo,
+            SnapshotShardSizeInfo.EMPTY,
+            randomLong()
+        );
+
+        // This should move a shard in an attempt to resolve the hot-spot
+        balancedShardsAllocator.allocate(routingAllocation);
+
+        assertEquals(1, routingAllocation.routingNodes().node(overloadedNode.getId()).numberOfOwningShards());
+        assertEquals(3, routingAllocation.routingNodes().node(otherNode.getId()).numberOfOwningShards());
+
+        final var clusterInfoSimulator = new ClusterInfoSimulator(routingAllocation);
+        final var movedShards = new HashSet<ShardRouting>();
+        for (RoutingNode routingNode : routingAllocation.routingNodes()) {
+            movedShards.addAll(routingNode.shardsWithState(ShardRoutingState.INITIALIZING).collect(Collectors.toSet()));
+        }
+        movedShards.forEach(shardRouting -> {
+            routingAllocation.routingNodes().startShard(shardRouting, new RoutingChangesObserver() {
+            }, randomNonNegativeLong());
+            clusterInfoSimulator.simulateShardStarted(shardRouting);
+        });
```

Comment:
I did this balance/simulate/balance sequence manually because it seemed a pain to set up the desiredBalanceAllocator, but I'd be open to doing that if we think this is worthwhile to make it a more realistic test.